packages/next/shared/lib/router/router.ts

```diff
@@ -492,7 +495,10 @@ function fetchNextData({
                * mapped location.
                * TODO: Change the status code in the handler.
                */
-              if (hasMiddleware && [301, 302, 308].includes(response.status)) {
+              if (
+                hasMiddleware &&
+                [301, 302, 307, 308].includes(response.status)
+              ) {
                 return { dataHref, response, text, json: {} }
```

Comment:
We've also have a fallback after this to not fail on 404 that we can remove now that we handle every data request!

Reply:
I think we may still need to handle the 404 case there if a user has an external rewrite that returns a 404 which we can't control